File: app/src/main/java/com/marvelcomics/brito/marvelcomics/ui/activity/home/HomeActivity.java
package com.marvelcomics.brito.marvelcomics.ui.activity.home;

import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.SearchView;
import android.view.View;

import com.marvelcomics.brito.entity.CharacterEntity;
import com.marvelcomics.brito.infrastructure.utils.AlertDialogUtils;
import com.marvelcomics.brito.marvelcomics.R;
import com.marvelcomics.brito.marvelcomics.databinding.ActivityHomeBinding;
import com.marvelcomics.brito.marvelcomics.ui.fragment.character.CharacterFragment;
import com.marvelcomics.brito.marvelcomics.ui.fragment.comics.ComicsFragment;
import com.marvelcomics.brito.marvelcomics.ui.fragment.series.SeriesFragment;
import com.marvelcomics.brito.presentation.presenter.home.HomeContract;
import com.marvelcomics.brito.presentation.presenter.home.HomePresenter;

import javax.inject.Inject;

import dagger.android.AndroidInjection;

public class HomeActivity extends AppCompatActivity implements HomeContract.View {

    private ActivityHomeBinding binding;

    @Inject
    protected HomePresenter homePresenter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        AndroidInjection.inject(this);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_home);
        binding.setViewActivity(this);
        homePresenter.setView(this);
        initListeners();
    }

    @Override
    public void showCharacter(CharacterEntity character) {
        binding.imageviewLoading.setVisibility(View.GONE);
        instantiateCharacterFragment(character);
        instantiateComicsFragment(character.getId());
        instantiateSeriesFragment(character.getId());
    }

    @Override
    public void showError(String message) {
        binding.imageviewLoading.setVisibility(View.GONE);
        AlertDialogUtils.showSimpleDialog("Erro", message, this);
    }

    public void searchViewFocus() {
        binding.searchviewMarvelCharacter.setIconified(false);
    }

    private void initListeners() {
        binding.searchviewMarvelCharacter.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                binding.imageviewLoading.setVisibility(View.VISIBLE);
                homePresenter.loadCharacter(query);
                return false;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                return false;
            }
        });
    }

    private void instantiateCharacterFragment(CharacterEntity characterEntity) {
        CharacterFragment fragment = CharacterFragment.newInstance(characterEntity);
        commitFragment(R.id.fragment_home_character, fragment);
    }

    private void instantiateComicsFragment(int id) {
        ComicsFragment fragment = ComicsFragment.newInstance(id);
        commitFragment(R.id.fragment_home_comics, fragment);
    }

    private void instantiateSeriesFragment(int id) {
        SeriesFragment fragment = SeriesFragment.newInstance(id);
        commitFragment(R.id.fragment_home_series, fragment);
    }

    private void commitFragment(int containerId, Fragment fragment) {
        getSupportFragmentManager().beginTransaction().replace(containerId,
                fragment, fragment.getClass().getSimpleName()).commit();
    }
}

File: app/src/main/java/com/marvelcomics/brito/view/HomeActivity.kt
package com.marvelcomics.brito.view

import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.marvelcomics.brito.databinding.ActivityMainBinding
import com.marvelcomics.brito.domain.entity.CharacterEntity
import com.marvelcomics.brito.hideKeyboard
import com.marvelcomics.brito.presentation.CharacterUiState
import com.marvelcomics.brito.presentation.GlobalUiState
import com.marvelcomics.brito.presentation.character.CharacterViewModel
import com.marvelcomics.brito.replaceFragment
import com.marvelcomics.brito.view.extensions.viewBinding
import com.marvelcomics.brito.view.fragment.character.CharacterFragment
import com.marvelcomics.brito.view.fragment.comics.ComicsFragment
import com.marvelcomics.brito.view.fragment.series.SeriesFragment
import kotlinx.coroutines.InternalCoroutinesApi
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch
import org.koin.androidx.viewmodel.ext.android.viewModel

@InternalCoroutinesApi
class HomeActivity : AppCompatActivity() {

    private val binding by viewBinding(ActivityMainBinding::inflate)
    private val characterViewModel: CharacterViewModel by viewModel()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(binding.root)
        initUi()
    }

    override fun onStart() {
        super.onStart()
        initObservers()
    }

    private fun initUi() {
        binding.buttonSearchMarvelCharacter.setOnClickListener {
            binding.edittextMarvelCharacter.text.toString().apply {
                if (this.isNotBlank()) {
                    getCharacterNav(this)
                    hideKeyboard()
                }
            }
        }
    }

    private fun initObservers() {
        lifecycleScope.launch {
            lifecycle.repeatOnLifecycle(Lifecycle.State.STARTED) {
                characterViewModel.characterUiState.collect {
                    when (it) {
                        is GlobalUiState.Loading -> {
                            binding.progressbarLoadingCharacter.visibility = View.VISIBLE
                        }
                        is GlobalUiState.NetworkError -> {
                            // do nothing
                        }
                        is CharacterUiState.Success -> {
                            binding.progressbarLoadingCharacter.visibility = View.GONE
                            val character = it.data as CharacterEntity
                            val characterFragment = CharacterFragment.newInstance(character)
                            replaceFragment(characterFragment, binding.fragmentHomeCharacter.id)
                            val comicsFragment = ComicsFragment.newInstance(character.id)
                            replaceFragment(comicsFragment, binding.fragmentHomeComics.id)
                            val seriesFragment = SeriesFragment.newInstance(character.id)
                            replaceFragment(seriesFragment, binding.fragmentHomeSeries.id)
                        }
                        is CharacterUiState.Error -> {
                            Toast.makeText(
                                this@HomeActivity,
                                "Error: ${it.exception.message}",
                                Toast.LENGTH_LONG
                            ).show()
                            binding.progressbarLoadingCharacter.visibility = View.GONE
                        }
                        else -> {
                            // do nothing
                        }
                    }
                }
            }
        }
    }

    private fun getCharacterNav(name: String) {
        characterViewModel.loadCharacter(name)
    }
}
